src/source/channel_router.rs

```diff
@@ -0,0 +1,304 @@
+        let mut new_map = channel_map.clone();
+        self.prepare_map(&mut new_map)?;
+        self.sender
+            .send(new_map)
+            .map_err(|_| ChannelMixerError::SendError)
+    }
+
+    fn prepare_map(&self, new_channel_map: &mut ChannelMap) -> Result<(), ChannelMixerError> {
+        if !new_channel_map
+            .iter()
+            .all(|(_from, to, _gain)| to < &self.out_channels_count)
+        {
+            return Err(ChannelMixerError::ConfigError);
+        }
+        new_channel_map.retain(|(_from, _to, gain)| *gain != 0.0);
+        new_channel_map.sort_by(|a, b| a.0.cmp(&b.0));
+        Ok(())
+    }
+}
+
+/// A source for extracting, reordering, mixing and duplicating audio between
+/// channels.
+// #[derive(Debug)] // TODO Reimplement debug? A Cell is not Debug.
+pub struct ChannelMixerSource<I>
+where
+    I: Source,
+    I::Item: Sample,
+{
+    /// Input [`Source`]
+    input: I,
+
+    /// Mapping of input to output channels.
+    channel_map: ChannelMap,
+
+    /// The output channel that [`next()`] will return next.
+    current_channel: u16,
+
+    /// The number of output channels.
+    channel_count: u16,
+
+    /// Helps to reduce dynamic allocation.
+    input_frame: Cell<Vec<I::Item>>,
+
+    /// The current input audio frame.
+    output_frame: Vec<Option<I::Item>>,
+
+    /// Communication channel with the controller.
+    receiver: Receiver<ChannelMap>,
+}
+
+impl<I> ChannelMixerSource<I>
+where
+    I: Source,
+    I::Item: Sample,
+{
+    /// Destroys this router and returns the underlying source.
+    #[inline]
+    pub fn into_inner(self) -> I {
+        self.input
+    }
+
+    /// Get mutable access to the inner source.
+    #[inline]
+    pub fn inner_mut(&mut self) -> &mut I {
+        &mut self.input
+    }
+}
+
+impl<I> Source for ChannelMixerSource<I>
+where
+    I: Source,
+    I::Item: Sample,
+{
+    #[inline]
+    fn current_span_len(&self) -> Option<usize> {
+        self.input.current_span_len()
+    }
+
+    #[inline]
+    fn channels(&self) -> u16 {
+        self.channel_count
+    }
+
+    #[inline]
+    fn sample_rate(&self) -> u32 {
+        self.input.sample_rate()
+    }
+
+    #[inline]
+    fn total_duration(&self) -> Option<std::time::Duration> {
+        self.input.total_duration()
+    }
+}
+
+impl<I> Iterator for ChannelMixerSource<I>
+where
+    I: Source,
+    I::Item: Sample,
+{
+    type Item = I::Item;
+
+    #[inline]
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.current_channel >= self.channel_count {
+            // TODO One may want to change mapping when incoming channel count changes.
+            if let Some(map_update) = self.receiver.try_iter().last() {
+                self.channel_map = map_update;
+            }
+
+            self.current_channel = 0;
+            self.output_frame.fill(None);
+            let input_channels = self.input.channels() as usize;
+
+            let mut input_frame = self.input_frame.take();
```

Comment:
you can use `std::mem::take` (replaces with Default) here and drop the `Cell` 

Then just an assignment where you do replace. 

I seem to be missing something however, why is it not operating on `self.input_frame` directly?

Reply:
The loop tries to get a frame from input's iterator and fill in the output frame at once. I did not have a better idea. It looks like `let input = &mut self.input;` helps borrow checker to resolve this (it now understands that only one field is borrowed allowing to mutate others). I have removed the `Cell` field altogether..